flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeParser.java

```diff
@@ -0,0 +1,916 @@
+				if (hasNextToken(TokenType.LIST_SEPARATOR)) {
+					nextToken(TokenType.LIST_SEPARATOR);
+					nextToken(TokenType.LITERAL_INT);
+					scale = tokenAsInt();
+				}
+				nextToken(TokenType.END_PARAMETER);
+			}
+			return new DecimalType(precision, scale);
+		}
+
+		private LogicalType parseDoubleType() {
+			if (hasNextToken(Keyword.PRECISION)) {
+				nextToken(Keyword.PRECISION);
+			}
+			return new DoubleType();
+		}
+
+		private LogicalType parseTimeType() {
+			int precision = TimeType.DEFAULT_PRECISION;
+			if (hasNextToken(TokenType.BEGIN_PARAMETER)) {
+				nextToken(TokenType.BEGIN_PARAMETER);
+				nextToken(TokenType.LITERAL_INT);
+				precision = tokenAsInt();
+				nextToken(TokenType.END_PARAMETER);
+			}
+			if (hasNextToken(Keyword.WITHOUT)) {
+				nextToken(Keyword.WITHOUT);
+				nextToken(Keyword.TIME);
+				nextToken(Keyword.ZONE);
+			}
+			return new TimeType(precision);
+		}
+
+		private LogicalType parseTimestampType() {
+			int precision = TimestampType.DEFAULT_PRECISION;
+			if (hasNextToken(TokenType.BEGIN_PARAMETER)) {
+				nextToken(TokenType.BEGIN_PARAMETER);
+				nextToken(TokenType.LITERAL_INT);
+				precision = tokenAsInt();
+				nextToken(TokenType.END_PARAMETER);
+			}
+			if (hasNextToken(Keyword.WITHOUT)) {
+				nextToken(Keyword.WITHOUT);
+				nextToken(Keyword.TIME);
+				nextToken(Keyword.ZONE);
+			} else if (hasNextToken(Keyword.WITH)) {
+				nextToken(Keyword.WITH);
+				if (hasNextToken(Keyword.LOCAL)) {
+					nextToken(Keyword.LOCAL);
+					nextToken(Keyword.TIME);
+					nextToken(Keyword.ZONE);
+					return new LocalZonedTimestampType(precision);
+				} else {
+					nextToken(Keyword.TIME);
+					nextToken(Keyword.ZONE);
+					return new ZonedTimestampType(precision);
+				}
+			}
+			return new TimestampType(precision);
+		}
+
+		private LogicalType parseIntervalType() {
+			nextToken(TokenType.KEYWORD);
+			switch (tokenAsKeyword()) {
+				case YEAR:
+				case MONTH:
+					return parseYearMonthIntervalType();
+				case DAY:
+				case HOUR:
+				case MINUTE:
+				case SECOND:
+					return parseDayTimeIntervalType();
+				default:
+					throw parsingError("Invalid interval resolution.");
+			}
+		}
+
+		private LogicalType parseYearMonthIntervalType() {
+			int yearPrecision = YearMonthIntervalType.DEFAULT_PRECISION;
+			switch (tokenAsKeyword()) {
+				case YEAR:
+					yearPrecision = parseOptionalPrecision(yearPrecision);
+					if (hasNextToken(Keyword.TO)) {
+						nextToken(Keyword.TO);
+						nextToken(Keyword.MONTH);
+						return new YearMonthIntervalType(
+							YearMonthResolution.YEAR_TO_MONTH,
+							yearPrecision);
+					}
+					return new YearMonthIntervalType(
+						YearMonthResolution.YEAR,
+						yearPrecision);
+				case MONTH:
+					return new YearMonthIntervalType(
+						YearMonthResolution.MONTH,
+						yearPrecision);
+				default:
+					throw parsingError("Invalid year-month interval resolution.");
+			}
+		}
+
+		private LogicalType parseDayTimeIntervalType() {
```

Comment:
How about splitting this method by the 3 cases.

Reply:
Moreover I think this method could be improved, if we first parse lower/upper resolution and then reuse `DataTypes.Resolution#resolveInterval`. In a pseudo code it could look sth like:

```
public Resolution parseAsResolution(IntervalUnit type) {
    if (type == DAY) {
	dayPrecision = parseOptionalPrecision(dayPrecision);
        return new Resolution(type, dayPrecision);
    } else if (lower == SECOND) {
	fractionalPrecision = parseOptionalPrecision(fractionalPrecision);
        return new Resolution(type, fractionalPrecision);
    }
}

public LogicalType parseDayTimeIntervalType() {
    IntervalUnit upper = tokenAsIntervalUnit();
    Optional<IntervalUnit> lower = tokenAsIntervalUnit();

    DataTypes.Resolution.resolveInterval(parseAsResolution(upper), lower.map(parseAsResolution).orNull)
}
```

Reply:
TBH I think it is not worth the effort. Usually, such classes are generated and look much worse. Unless, the standard changes we will also not need to make changes in this method. So maintainability is also not an issue. Exposing `DataTypes.Resolution.resolveInterval` is mixing API with internal utils.